### ISSUE TITLE
Provide a way to enable both quantity_support and time_support in one go

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#8860]
In https://github.com/astropy/astropy/pull/8782 we discussed that it would be nice to have a way to enable both ``time_support`` and ``quantity_support``, maybe e.g. ``astropy_support``.

**Comment by neutrinoceros** [comment]
@spranav1205 Hi, thanks for looking into this !

Nowadays Python's standard lib is richer and I think we can go with a simpler desgin that the one that was choosen at the time for `time_support` and `quantity_support`.
In particular, for this specific task of combining existing context managers, I think we chould use (or take inspiration from) [`contextlib.ExitStack`](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack).

**Comment by spranav1205** [comment]
We can use the following approach to combine `quantity_support` and `time_support` using a function, without the need for a class:

    @contextmanager
    def astro_support():
     
        q = quantity_support()
        t = time_support()
        
        try:
            q.__enter__()
            t.__enter__()
            yield  
        finally:
            q.__exit__()
            t.__exit__()

With this implementation, astro_support can be used as a context manager with the `with` statement, but it can't be used independently like` quantity_support()` or `time_support()`.

Alternatively, we can implement `astro_support` as a class, which can act both as a context manager (with `with`) and as a decorator for functions. This would look something like:

    class astro_support(ContextDecorator):
              def __init__(self):
                  # Call both quantity and time support functions and assign them to a class variable, e.g., self.q_support
                 
              def __enter__(self):
                  # Enter both quantity and time support context
                  
              def __exit__(self, exc_type, exc_value, traceback):
                  # Exit both quantity and time support context
                  
Would you like me to implement either of these? Or should we explore some other method?

**Comment by neutrinoceros** [comment]
Actually the `contextlib.contexmanager` implementation also allows using it as a decorator

quoting from https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager
> [contextmanager()](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) uses [ContextDecorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) so the context managers it creates can be used as decorators as well as in [with](https://docs.python.org/3/reference/compound_stmts.html#with) statements. When used as a decorator, a new generator instance is implicitly created on each function call (this allows the otherwise “one-shot” context managers created by [contextmanager()](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) to meet the requirement that context managers support multiple invocations in order to be used as decorators).

So my preference is to go with the first approach as it's simpler. Note that it is generally not recommended to call so-called dunder[^1] methods like `__enter__` and `__exit__` directly, so I would favor something like:

```python
from contextlib import contextmanager, ExitStack

@contextmanager
def astro_support():
	with ExitStack() as stack:
		stack.enter_context(quantity_support())
        stack.enter_context(time_support())
        yield
```

[^1]: "dunder" is short for "double underscore" and is a common way to refer to Python's protocols that are named with two leading and two trailing underscores

**Comment by spranav1205** [comment]
Yes, this works perfectly. I have opened a PR #17575. I will update the changelog and add a test case shortly!

**Comment by Subham-KRLX** [comment]
Hi @astropy/core team 👋
I'd love to help close out issue #8860 (“enable both quantity_support and time_support in one go”). I saw PR #17979 is already addressing this great start!
I can help with:

Adding unit tests (both context manager and decorator use cases)
Updating the visualization documentation
Would it be helpful if I take this on next?

**Comment by neutrinoceros** [comment]
Thanks for pointing out #17979, which we apparently missed entirely. We'd actually prefer to get both code and tests in the same PR, so I'd like to see if @elmostafa-abdelaal is still interested in this after month of not hearing back from us. In case there's no response (which would be understandable given the circumstances) after a reasonable amount of time (say a couple weeks at least), you (@Subham-KRLX) would be welcome to take this on, possibly starting from where things are now in #17979 and adding a test (but please don't forget to keep @elmostafa-abdelaal as a commit author if you do !). Thanks for your interest and patience.

**Comment by Subham-KRLX** [comment]
Thanks for the update I totally understand waiting to hear back from @elmostafa-abdelaal it makes sense to give them first dibs since they’ve already put work into #17979.
If it’s helpful I am happy to:
Help draft tests/docs in the meantime (maybe start a WIP branch?),Jump in later if needed while crediting their commits or Just review whenever the PR moves forward.
Whatever works best for the team just let me know.

**Comment by neutrinoceros** [comment]
I think it'll be easier for everyone if you wait a bit before jumping in. In the meantime, maybe drafting test cases could be done as a PR on your fork (not to be merged, but a comfortable setup to see a diff and have a conversation if needed) ?

**Comment by adityapathak-dev** [comment]
i would like to help fix this

**Comment by LoveChauhan-18** [comment]
Hello, I would like to contribute to this issue. Would it be possible to assign it to me?

**Comment by pllim** [comment]
How do people feel about the proposed implementation at https://github.com/astropy/astropy/pull/19061 ?

**Comment by elmostafa-abdelaal** [comment]
Okay

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by molly22-byte** [comment]
Hello, I would like to contribute to this issue. Would it be possible to assign it to me? OR can you please suggest me other issues to work upon?

**Comment by thiagodcoq** [comment]
Hey, can I work on this issue?
